rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py

```diff
@@ -341,72 +363,255 @@ def _convert_attribute_tokens_to_texts(
 
         return attribute_texts
 
-    def _train_with_shared_vocab(self, attribute_texts: Dict[Text, List[Text]]):
-        """Construct the vectorizers and train them with a shared vocab"""
-
-        self.vectorizers = self._create_shared_vocab_vectorizers(
-            {
-                "strip_accents": self.strip_accents,
-                "lowercase": self.lowercase,
-                "stop_words": self.stop_words,
-                "min_ngram": self.min_ngram,
-                "max_ngram": self.max_ngram,
-                "max_df": self.max_df,
-                "min_df": self.min_df,
-                "max_features": self.max_features,
-                "analyzer": self.analyzer,
-            }
-        )
+    @staticmethod
+    def _get_starting_empty_index(vocabulary: Dict[Text, int]) -> int:
+        for key in vocabulary.keys():
+            if key.startswith(BUFFER_SLOTS_PREFIX):
+                return int(key.split(BUFFER_SLOTS_PREFIX)[1])
+        return len(vocabulary)
+
+    def _update_vectorizer_vocabulary(
+        self, attribute: Text, new_vocabulary: Set[Text]
+    ) -> None:
+        """Update the existing vocabulary of the vectorizer with new unseen words.
 
-        combined_cleaned_texts = []
-        for attribute in self._attributes:
-            combined_cleaned_texts += attribute_texts[attribute]
+        These unseen words should only occupy the empty buffer slots.
 
-        try:
-            self.vectorizers[TEXT].fit(combined_cleaned_texts)
-        except ValueError:
+        Args:
+            attribute: Message attribute for which vocabulary should be updated.
+            new_vocabulary: Set of words to expand the vocabulary with if they are unseen.
+        """
+        existing_vocabulary: Dict[Text, int] = self.vectorizers[attribute].vocabulary
+        if len(new_vocabulary) > len(existing_vocabulary):
             logger.warning(
```

Comment:
sure this shouldn't be a `warnings.warn`?

Reply:
Changed to make use of `rasa.shared.utils.io.raise_warning`.